src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -479,357 +487,304 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
+                        // frozen here.
+                        var inProgressState = InProgressState.Create(
+                            isFrozen: false, compilation, CompilationTrackerGeneratorInfo.Empty, staleCompilationWithGeneratedDocuments: null,
+                            ImmutableList<(ProjectState, CompilationAndGeneratorDriverTranslationAction)>.Empty);
+                        WriteState(inProgressState);
+                        return inProgressState;
+                    }
+                    catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
                     {
-                        try
-                        {
-                            var compilation = CreateEmptyCompilation();
-
-                            using var _ = ArrayBuilder<SyntaxTree>.GetInstance(ProjectState.DocumentStates.Count, out var trees);
-                            foreach (var documentState in ProjectState.DocumentStates.GetStatesInCompilationOrder())
-                            {
-                                cancellationToken.ThrowIfCancellationRequested();
-                                // Include the tree even if the content of the document failed to load.
-                                trees.Add(await documentState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false));
-                            }
-
-                            compilation = compilation.AddSyntaxTrees(trees);
-
-                            // We only got here when we had no compilation state at all.  So we couldn't have gotten
-                            // here from a frozen state (as a frozen state always ensures we have a
-                            // WithCompilationTrackerState).  As such, we can safely still preserve that we're not
-                            // frozen here.
-                            var allSyntaxTreesParsedState = new AllSyntaxTreesParsedState(
-                                isFrozen: false, compilation, CompilationTrackerGeneratorInfo.Empty, compilationWithGeneratedDocuments: null);
-                            WriteState(allSyntaxTreesParsedState);
-                            return allSyntaxTreesParsedState;
-                        }
-                        catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
-                        {
-                            throw ExceptionUtilities.Unreachable();
-                        }
+                        throw ExceptionUtilities.Unreachable();
                     }
+                }
 
-                    async Task<FinalState> BuildFinalStateFromInProgressStateAsync(
-                        SolutionCompilationState compilationState, InProgressState inProgressState, CancellationToken cancellationToken)
+                async Task<InProgressState> CollapseInProgressStateAsync(InProgressState initialState)
```

Comment:
this logic became much simpler.  just a loop that keeps on popping off translation actions, producing the next inprogresstate.